v1_1/system.ts

```diff
@@ -0,0 +1,167 @@
+import {CryptoV1_1, ILocationData, IVisit, tr} from "./crypto";
+import {crypto_box_seal, crypto_box_seal_open, from_base64, IKeyPair, randombytes_buf, to_base64} from "../lib/sodium";
+import {QRCodeContent, QRCodeTrace} from "./proto";
+
+export interface ITrace {
+    tr: tr;
+    start: number;
+    end: number;
+}
+
+/**
+ * The System package uses the crypto but only passes around base64 encoded protobuf messages. This helps to
+ * understand which fields need to be serialized / deserialized.
+ */
+
+/**
+ * HealthAuthority is responsible to keep the private key and to publish the traces of
+ * locations that had a positively tested visitor.
+ */
+export class HealthAuthority {
+    public keyPair: IKeyPair;
+
+    constructor() {
+        this.keyPair = CryptoV1_1.setupHA();
+    }
+
+    /**
+     * Takes the qrTrace from a location, decrypts it, and returns the trace that needs to be
+     * sent to all visitors.
+     *
+     * @param qrTrace the qr-code from the location
+     * @param info about the location
+     * @param start in seconds since unix epoch
+     * @param end in seconds since unix epoch
+     */
+    createTrace(qrTrace: string, info: string, start: number, end: number): ITrace {
+        const trace = qrTrace.replace(/^.*#/, '');
+        const qrBuf = crypto_box_seal_open(
+            from_base64(trace),
+            this.keyPair.publicKey,
+            this.keyPair.privateKey
+        );
+        const qrCodeTrace = QRCodeTrace.decode(qrBuf);
+        if (qrCodeTrace.info !== info) {
+            throw new Error("not the same info");
+        }
+        const tr = {sk: qrCodeTrace.trSk, r2: qrCodeTrace.trR2};
+        if (!CryptoV1_1.verifyTrace(info, tr,
+            {
+                r1: qrCodeTrace.pTrR1,
+                r2: qrCodeTrace.pTrR2,
+            })
+        ) {
+            throw new Error("trace verification fails")
+        }
+        return {tr, start, end}
+    }
+}
+
+/**
+ * Location is used by a location owner to create the two QRCodes.
+ */
+export class Location {
+    public data: ILocationData;
+    public notificationKey = randombytes_buf(32);
+
+    constructor(public healthAuthorityPubKey: Uint8Array,
+                public name: string,
+                public location: string,
+                public room: string,
+                public locationType: number) {
+        this.data = CryptoV1_1.genCode(this.infoStr());
+    }
+
+    static infoStr(name: string, location: string, room: string, locationType: number): string {
```

Comment:
This needs to be something much more robust. And it ignores `notification_key`. This should really do something similar to `SeedMessage` and then messagepacking that.

Reply:
replaced with InfoLocation which is a protobuf